File: SDUI/src/sdui/normalize.py
"""Resolve component uses into bounded instance trees, independent of GUI toolkits."""
from dataclasses import dataclass, replace
from .ast import Literal, Reference, Span
from .formatting import check_combination, fail, formatting
from .validate import widget_arguments


@dataclass(frozen=True)
class Instance:
    kind: str
    path: str
    variant: str | None
    widget: str | None
    text: str | None
    arguments: tuple[tuple[str, Literal | Reference], ...]
    layout: tuple[tuple[str, object], ...]
    rows: tuple[tuple['Instance', ...], ...]
    regions: tuple[tuple[str, 'Instance'], ...]
    span: Span


def resolve_all(document):
    definitions = {definition.name: definition.root for definition in document.definitions}
    count = 0

    def expand(node, path, depth):
        nonlocal count
        count += 1
        if count > 8192:
            fail('expansion-limit', 'More than 8192 expanded components', node)
        if depth > 64:
            fail('depth-limit', 'Expanded component depth exceeds 64', node)
        if node.kind == 'use':
            if node.target not in definitions:
                fail('unknown-definition', f'Unknown component {node.target}', node)
            instance = expand(definitions[node.target], path, depth + 1)
            # Validate suffix against the effective component, not the use-site node.
            effective = replace(node, kind=instance.kind,
                                rows=instance.rows)
            overlay = formatting(effective)
            props = dict(instance.layout)
            props.update(overlay)
            check_combination(props, instance.kind, effective)
            check_wrap_children(props, instance.rows, node)
            return replace(instance, layout=tuple(props.items()))
        props = formatting(node)
        rows, regions = [], []
        for r, row in enumerate(node.rows):
            items = []
            for c, child in enumerate(row.items):
                child_name = child.role or child.name or child.target
                if child_name:
                    child_path = path + '/' + child_name
                else:
                    child_path = f'{path}/$r{r}c{c}'
                instance = expand(child, child_path, depth + 1)
                if node.kind == 'group' and instance.kind == 'frame':
                    fail('group-content', 'Frame reference cannot be placed in a widget group', child)
                if child.role:
                    regions.append((child.role, instance))
                else:
                    items.append(instance)
            if items:
                rows.append(tuple(items))
        check_wrap_children(props, rows, node)
        return Instance(node.kind, path, 'box' if node.variant else None,
                        node.widget, node.text.value if node.text else None,
                        tuple(widget_arguments(node).items()) if node.kind == 'widget' else (),
                        tuple(props.items()), tuple(rows), tuple(regions), node.span)
    return {name: expand(node, name, 1) for name, node in definitions.items()}


def normalize(document):
    from .validate import validate
    validate(document)
    return resolve_all(document)


def check_wrap_children(props, rows, node):
    if props.get('wrap') == 'wrap':
        for row in rows:
            for child in row:
                x = dict(child.layout).get('x', '')
                if x == 'fill' or str(x).endswith('fr'):
                    fail('wrap-layout', 'Wrapped children cannot use horizontal fill/fr', node)

File: SDUI/go/parser/normalize.go
package parser

import (
	"fmt"
	"strings"
)

type Region struct {
	Role string    `json:"role"`
	Node *Instance `json:"node"`
}
type Instance struct {
	Kind      string         `json:"kind"`
	Path      string         `json:"path"`
	Variant   string         `json:"variant,omitempty"`
	Widget    string         `json:"widget,omitempty"`
	Text      string         `json:"text,omitempty"`
	Arguments map[string]any `json:"arguments"`
	Layout    map[string]any `json:"layout"`
	Rows      [][]*Instance  `json:"rows"`
	Regions   []Region       `json:"regions"`
	Span      Span           `json:"span"`
}

func (i *Instance) Region(role string) *Instance {
	for _, r := range i.Regions {
		if r.Role == role {
			return r.Node
		}
	}
	return nil
}
func (i *Instance) Walk(fn func(*Instance)) {
	fn(i)
	for _, r := range i.Regions {
		r.Node.Walk(fn)
	}
	for _, row := range i.Rows {
		for _, c := range row {
			c.Walk(fn)
		}
	}
}
func (i *Instance) Argument(name string) string {
	if v, ok := i.Arguments[name].(Literal); ok {
		if text, ok := v.Value.(string); ok {
			return text
		}
	}
	return ""
}
func checkWrap(p map[string]any, rows [][]*Instance, s Span) {
	if p["wrap"] == "wrap" {
		for _, row := range rows {
			for _, child := range row {
				x := child.Layout["x"]
				if x == "fill" || strings.HasSuffix(fmt.Sprint(x), "fr") {
					fail("wrap-layout", "Wrapped child has fill/fr", s)
				}
			}
		}
	}
}

// Normalize returns bounded, independent instance trees. Mutating a tree never changes the AST.
func Normalize(d *Document) (roots map[string]*Instance, err error) {
	defer recoverDiagnostic(&err)
	validateLocal(d)
	defs := map[string]*Node{}
	for _, def := range d.Definitions {
		defs[def.Name] = def.Root
	}
	count := 0
	var expand func(*Node, string, int) *Instance
	expand = func(n *Node, path string, depth int) *Instance {
		count++
		if count > 8192 {
			fail("expansion-limit", "More than 8192 expanded components", n.Span)
		}
		if depth > 64 {
			fail("depth-limit", "Expanded depth exceeds 64", n.Span)
		}
		if n.Kind == "use" {
			i := expand(defs[val(n.Target)], path, depth+1)
			overlay := formatting(n, i.Kind, len(i.Rows))
			for k, v := range overlay {
				i.Layout[k] = v
			}
			checkCombination(i.Layout, i.Kind, len(i.Rows), n.Span)
			checkWrap(i.Layout, i.Rows, n.Span)
			return i
		}
		props := formatting(n, n.Kind, len(n.Rows))
		i := &Instance{Kind: n.Kind, Path: path, Widget: val(n.Widget), Layout: props, Arguments: map[string]any{}, Rows: [][]*Instance{}, Regions: []Region{}, Span: n.Span}
		if n.Variant != nil {
			i.Variant = "box"
		}
		if n.Text != nil {
			i.Text = n.Text.Value.(string)
		}
		if n.Kind == "widget" {
			i.Arguments = widgetArguments(n)
		}
		for r, row := range n.Rows {
			items := []*Instance{}
			for c, child := range row.Items {
				name := val(child.Role)
				if name == "" {
					name = val(child.Name)
				}
				if name == "" {
					name = val(child.Target)
				}
				if name == "" {
					name = fmt.Sprintf("$r%dc%d", r, c)
				}
				instance := expand(child, path+"/"+name, depth+1)
				if n.Kind == "group" && instance.Kind == "frame" {
					fail("group-content", "Frame reference in widget group", child.Span)
				}
				if child.Role != nil {
					i.Regions = append(i.Regions, Region{*child.Role, instance})
				} else {
					items = append(items, instance)
				}
			}
			if len(items) > 0 {
				i.Rows = append(i.Rows, items)
			}
		}
		checkWrap(props, i.Rows, n.Span)
		return i
	}
	roots = map[string]*Instance{}
	for _, def := range d.Definitions {
		roots[def.Name] = expand(def.Root, def.Name, 1)
	}
	validateConnections(d, roots)
	return roots, nil
}

// Compile shares exactly the Parse/Normalize path used by all hosts.
func Compile(source string) (*Document, map[string]*Instance, error) {
	d, err := Parse(source)
	if err != nil {
		return nil, nil, err
	}
	roots, err := Normalize(d)
	if err != nil {
		return nil, nil, err
	}
	return d, roots, nil
}
